nighthawk/README.md

```diff
@@ -0,0 +1,154 @@
+On Ubuntu, run the following command:
+
+``` bash
+
+# For Nighthawk / Envoy. 
+# Realpath may give an error, when it does that you probably already have it.
+
+sudo apt-get install \
+   libtool \
+   cmake \
+   realpath \
+   clang-format-7 \
+   automake \
+   ninja-build \
+   curl \
+   unzip
+
+# For tools/stats.py
+sudo apt-get install python3 python3-pip
+sudo pip3 install hdrhistogram jsonpickle
+```
+
+## Building and testing Nighthawk
+```zsh
+# TODO(oschaaf): Collect and list prerequisites. 
+# Currently mostly the same as Envoy.
+
+# build it
+bazel build //:nighthawk_client
+
+# test it
+bazel test //test:nighthawk_test
+```
+
+## Using the nighthawk client
+
+```
+➜ bazel-bin/nighthawk_client --help
+
+USAGE:
+
+   bazel-bin/nighthawk_client  [-v <trace|debug|info|warn|error>]
+                               [--concurrency <string>] [--h2] [--timeout
+                               <uint64_t>] [--duration <uint64_t>]
+                               [--connections <uint64_t>] [--rps
+                               <uint64_t>] [--] [--version] [-h] <uri
+                               format>
+
+
+Where:
+
+   -v <trace|debug|info|warn|error>,  --verbosity <trace|debug|info|warn
+      |error>
+     Verbosity of the output. Possible values: [trace, debug, info, warn,
+     error, critical]. The default level is 'info'.
+
+   --concurrency <string>
+     The number of concurrent event loops that should be used. Specify
+     'auto' to let nighthawk run leverage all (aligned) vCPUs. Note that
+     increasing this effectively multiplies configured --rps and
+     --connection values. Default: 1.
+
+   --h2
+     Use HTTP/2
+
+   --timeout <uint64_t>
+     Timeout period in seconds used for both connection timeout and grace
+     period waiting for lagging responses to come in after the test run is
+     done. Default: 5.
+
+   --duration <uint64_t>
+     The number of seconds that the test should run. Default: 5.
+
+   --connections <uint64_t>
+     The number of connections that the test should maximally use. Default:
+     1.
+
+   --rps <uint64_t>
+     The target requests-per-second rate. Default: 5.
+
+   --,  --ignore_rest
+     Ignores the rest of the labeled arguments following this flag.
+
+   --version
+     Displays version information and exits.
+
+   -h,  --help
+     Displays usage information and exits.
+
+   <uri format>
+     (required)  uri to benchmark. http:// and https:// are supported, but
+     in case of https no certificates are validated.
+
+
+   Nighthawk is a web server benchmarking tool.
+```
+
+## Sample benchmark run
+
+```zsh
+# start envoy on core 2
+$ taskset -c 3 /path/to/envoy --config-path tools/envoy.yaml
+
+# run the benchmark on cores 0 and 1
+$ taskset -c 0-1 bazel-bin/nighthawk_client --concurrency auto --rps 30000 --connections 1 --duration 3 http://127.0.0.1:10000/ && tools/stats.py res.txt benchmark
```

Comment:
Should we have a section documenting benchmarking best practices? E.g. using tasksets, disabling power management, considerations when running in VMs, using a dedicated machine, understanding variance and statistical significance, using opt builds, Envoy config settings that make sense for benchmarking, etc.